internal/clients/kafka/acl/acl.go

```diff
@@ -0,0 +1,187 @@
+package acl
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/crossplane-contrib/provider-kafka/apis/acl/v1alpha1"
+
+	"github.com/pkg/errors"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kmsg"
+)
+
+// AccessControlList is a holistic representation of a Kafka ACL with configurable
+// fields
+type AccessControlList struct {
+	Name                      string `json:"Name"`
+	ResourceType              string `json:"ResourceType"`
+	Principle                 string `json:"Principle"`
+	Host                      string `json:"Host"`
+	Operation                 string `json:"Operation"`
+	PermissionType            string `json:"PermissionType"`
+	ResourcePatternTypeFilter string `json:"ResourcePatternTypeFilter"`
+}
+
+// List lists all the ACLs in Kafka
+func List(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) (*AccessControlList, error) {
+
+	o, err := kmsg.ParseACLOperation(strings.ToLower(accessControlList.Operation))
+	if err != nil {
+		return nil, errors.Wrap(err, "did not return ACL Operation")
+	}
+
+	ao := []kadm.ACLOperation{o}
+
+	rpt, err := kmsg.ParseACLResourcePatternType(strings.ToLower(accessControlList.ResourcePatternTypeFilter))
+	if err != nil {
+		return nil, errors.Wrap(err, "did not return parsing of ACL pattern")
+	}
+
+	b := kadm.ACLBuilder{}
+	ab := b.Topics(accessControlList.Name).Allow(accessControlList.Principle).AllowHosts(accessControlList.Host).Operations(ao[0]).ResourcePatternType(rpt)
+
+	resp, err := cl.DescribeACLs(ctx, ab)
+
+	if resp == nil {
+		return nil, errors.Wrap(err, "describe ACLs response is empty")
+	}
+
+	exists := resp[0].Described
+
+	if exists == nil {
+		return nil, errors.Wrap(err, "cannot describe ACL, it does not exist")
+	}
+
+	acl := AccessControlList{}
+	acl.ResourceType = accessControlList.ResourceType
+	acl.Principle = accessControlList.Principle
+	acl.Host = accessControlList.Host
+	acl.Operation = accessControlList.Operation
+	acl.PermissionType = accessControlList.PermissionType
+	acl.ResourcePatternTypeFilter = accessControlList.ResourcePatternTypeFilter
+
+	return &acl, nil
+}
+
+// Create creates an ACL from the Kafka side
+func Create(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) error {
+
+	o, _ := kmsg.ParseACLOperation(strings.ToLower(accessControlList.Operation))
+	ao := []kadm.ACLOperation{o}
+
+	rpt, _ := kmsg.ParseACLResourcePatternType(strings.ToLower(accessControlList.ResourcePatternTypeFilter))
+
+	b := kadm.ACLBuilder{}
+	ab := b.Allow(accessControlList.Principle).AllowHosts(accessControlList.Host).Operations(ao[0]).ResourcePatternType(rpt)
+
+	switch accessControlList.ResourceType {
+	case "Topic":
+		ab = ab.Topics(accessControlList.Name)
+	case "Group":
+		ab = ab.Groups(accessControlList.Name)
+	case "TransactionalID":
+		ab = ab.TransactionalIDs(accessControlList.Name)
```

Comment:
We went through adding support for Cluster and Any, but found in both cases that it broke our Kafka clusters. We checked with our Kafka group and found that they were not supporting resource types outside of Topic and made a decision to remove it. Do you think we should add this in even if it might break the clusters for our users?

Reply:
It is interesting that try to add an ACL for Cluster type to break Kafka clusters. I would expect more graceful handling.

Considering this provider to be a more generic Kafka provider I think we should support it. In the worst case, I can think of dropping `Any` and `Cluster` from enum list for the field.